utils/cpu_features.c

```diff
@@ -26,28 +32,119 @@ int instrset_detect(void)
+    iset = SSE2;                                           // 2: SSE2 supported
+
     if ((abcd[2] & (1 <<  0)) == 0) return iset;           // no SSE3
-    iset = 3;                                              // 3: SSE3 supported
+    iset = SSE3;                                           // 3: SSE3 supported
+
     if ((abcd[2] & (1 <<  9)) == 0) return iset;           // no SSSE3
-    iset = 4;                                              // 4: SSSE3 supported
+    iset = SSSE3;                                          // 4: SSSE3 supported
+
     if ((abcd[2] & (1 << 19)) == 0) return iset;           // no SSE4.1
-    iset = 5;                                              // 5: SSE4.1 supported
+    iset = SSE4;                                           // 5: SSE4.1 supported
+
     if ((abcd[2] & (1 << 23)) == 0) return iset;           // no POPCNT
     if ((abcd[2] & (1 << 20)) == 0) return iset;           // no SSE4.2
-    iset = 6;                                              // 6: SSE4.2 supported
+    iset = SSE42;                                          // 6: SSE4.2 supported
+
     if ((abcd[2] & (1 << 27)) == 0) return iset;           // no OSXSAVE
     if ((xgetbv(0) & 6) != 6)       return iset;           // AVX not enabled in O.S.
     if ((abcd[2] & (1 << 28)) == 0) return iset;           // no AVX
-    iset = 7;                                              // 7: AVX supported
+    iset = AVX;                                            // 7: AVX supported
+
     cpuid(abcd, 7);                                        // call cpuid leaf 7 for feature flags
     if ((abcd[1] & (1 <<  5)) == 0) return iset;           // no AVX2
-    iset = 8;                                              // 8: AVX2 supported
+    iset = AVX2;                                           // 8: AVX2 supported
+
     cpuid(abcd, 0xD);                                      // call cpuid leaf 0xD for feature flags
     if ((abcd[0] & 0x60) != 0x60)   return iset;           // no AVX512
-    iset = 9;                                              // 9: AVX512F supported
+    iset = AVX512F;                                        // 9: AVX512F supported
     return iset;
 }
 
+// Report the max ISA supported by both the CPU and compiler
+int get_max_usable_isa(void)
+{
+    static int iset = -1;                                  // remember value for next call
+    if (iset >= 0) {
+        return iset;                                       // called before
+    }
+    iset = runtime_instrset_detect();
+
+    switch(iset){
+        case AVX512F:
+#ifdef __AVX512F__
+            iset = AVX512F;
+            break;
+#elif defined(GAS_BUG_DISABLE_AVX512)
+            fprintf(stderr, "[Warning] AVX512 is disabled due to a GNU Assembler bug.  Upgrade to binutils >= 2.32 to fix this.\n");
+#else
+            fprintf(stderr, "[Warning] The CPU supports AVX512 but the compiler does not.  Can you try another compiler?\n");
```

Comment:
Nit-pick - the statements should be that "... AVX512F is disabled ..." or "... CPU supports AVX512F ..."